README.md
<p align="center">
<img width="200" src="https://img.icons8.com/clouds/500/000000/synchronize.png"/>
</P>
 <p align="center">
 <img src="https://img.shields.io/badge/coverage-100%25-brightgreen">
 <img src="https://img.shields.io/badge/min%20size-1%20kb-blue">
 <img src="https://img.shields.io/npm/dt/data-matcher.svg?colorB=ff69b4">
 </p>

## 数据适配器

> 本文档是 v2 版本内容，v1 版本已经处于废弃 ⚠️ 阶段，请及时升级（codeMod 正在开发中，当前请手动修改）。维护阶段如果仍需参考 v1 文档，请访问 [v1 文档](https://github.com/xiamu14/data-matcher/blob/master/src/v1/README.md)

## 用途

在数据处理过程中，不同环节所储存或反馈的数据字段、类型总会存在差异。

服务端 api 数据和前端数据的差异，前端组件和第三方组件数据的差异，数据在不同环节的差异等。

如果基于同一领域，同一场景，同一产品的认知，这种差异往往只是数据表象的差异，比如：key 命名不一致，value 数据类型不同，数据原始或复合状态不同。

通过开发经验归纳，大致有下面三种差异：

1. 初始数据和使用时数据的复合形态不同：初始数据应该包含产品所需的全部内容，但在使用时可以需要不同的复合形态，比如：select 选项初始数据包含 {id,name}，第三方组件使用时需要 {key,value,label}。[add ，delete 方法组合使用可抹平差异]
2. key 命名不一致 [editKey 方法可修改 key]
3. value 数据类型不一致 [editValue 方法可修改 value]

当然，实际开发中还存在更复杂的差异，那样的情状建议特例化处理。

## 安装

```bash
yarn add data-matcher
```

```bash
pnpm install data-matcher
```

## 使用

```js
const data = { a: 'a', b: 'b' };
const matcher = new Matcher(data);
matcher
  .add('c', () => 'c')
  .delete(['b'])
  .editValue('a', () => 'aa');
matcher.data; // { a: 'aa', c: 'c' }
```

## 方法

### add (增加数据)

- 定义

  ```js
  public add(key: DataItemKey, valueFn: (data: any) => any)
  ```

- 参数
  | 参数 | 描述 | 默认值 | 类型 |
  | ------ | ----------- | ------ | ------ |
  | key | 键 | -- | string 或 Symbol |
  | valueFn | 值的生成函数 | -- | (data: DataItem) => any |

- 示例
  ```js
  const data = { startTime: '2019/09/12', endTime: '2019/09/30' };
  const matcher = new Matcher(data);
  matcher.add('dateRange', (data) => `${data.startTime}-${data.endTime}`);
  matcher.data; // {startTime: '2019/09/12', endTime: '2019/09/30', dateRange:'2019/09/12-2019/09/30'}
  ```

### delete (删除数据)

- 定义

  ```js
  public delete(keys: DataItemKey[])
  ```

- 参数
  | 参数 | 描述 | 默认值 | 类型 |
  | ------ | ----------- | ------ | ------ |
  | keys | 键数组 | -- | 包含 string 或 Symbol 的数组 |

- 示例
  ```js
  const data = { key: '1', label: 'apple', value: 'apple' };
  const matcher = new Matcher(data);
  matcher.delete(['label']);
  matcher.data; // {key:'1',value:'apple'}
  ```

### editValue (修改 value)

- 定义

  ```js
  public editValue(
    key: DataItemKey,
    valueFn: (value: any, data: DataItem) => any,
  )
  ```

- 参数
  | 参数 | 描述 | 默认值 | 类型 |
  | ------ | ----------- | ------ | ------ |
  | key | 键 | -- | string 或 Symbol |
  | valueFn | 值的生成函数 | -- | (key:any, data: DataItem) => any |

- 示例
  ```js
  const data = { price: 1, createAt: 1632833413149 }; // 价格服务端存储单位[分]
  const matcher = new Matcher(data);
  matcher
    .editValue('price', (value) => value / 100)
    .editValue('createAt', (value) => dayjs(value).format('YYYY/MM/DD'));
  matcher.data; // {price:0.01, createAt:'2021/09/28'}
  ```

### editKey (修改 key)

- 定义

  ```js
  public editKey(keyMap: Record<DataItemKey, DataItemKey>)
  ```

- 参数
  | 参数 | 描述 | 默认值 | 类型 |
  | ------ | ----------- | ------ | ------ |
  | keyMap | 新旧 key 对象 | -- | {旧 key：新 key} |

- 示例
  ```js
  const data = { id: '1' };
  const matcher = new Matcher(data);
  matcher.editKey({ id: 'key' });
  marcher.data; // {key:'1'}
  ```

### clone (拷贝数据)

- 定义

  ```js
  public clone(keyMap: Record<DataItemKey, DataItemKey>)
  ```

- 参数
  | 参数 | 描述 | 默认值 | 类型 |
  | ------ | ----------- | ------ | ------ |
  | keyMap | key 对象 | -- | {元 key：新 key} |

- 示例
  ```js
  const data = { a: 'a', b: { bb: 'bb' } };
  const matcher = new Matcher(data);
  matcher.clone({ a: 'a/' });
  matcher.data; // {a:'a', b:{bb:'bb'}, 'a/':'a'}
  ```

### clean (清除无意义的数据)

- 定义

  ```js
  public clean(invalidValues: MayBeInvalidType[])
  ```

- 参数
  | 参数 | 描述 | 默认值 | 类型 |
  | ------ | ----------- | ------ | ------ |
  | invalidValues | 无意义的值集合 | -- | null 、 undefined 、 '' 、 'null' 、 'undefined' 、 0 |

- 示例

  ```js
  const data = { price: 100, projectId: undefined };
  const matcher = new Matcher(data);
  matcher.clean([undefined]);
  matcher.data; // {price:100}
  ```

### when (根据条件组合操作)

- 定义

  ```js
  public when(
    condition: boolean,
    whenTruthy: ((that: Matcher) => void) | null,
    whenFalsy: ((that: Matcher) => void) | null,
  )
  ```

- 参数
  | 参数 | 描述 | 默认值 | 类型 |
  | ------ | ----------- | ------ | ------ |
  | condition | 条件值 | -- | boolean | true or false |

- 示例
  ```js
  const data = { a: 'a', b: { bb: 'bb' } };
  const matcher = new Matcher(data);
  matcher.when(data.a === 'a', (that) => that.add('aa', () => 'aa'), null);
  matcher.data; // {a:'a', b:{bb:'bb'}, 'aa':'aa'}
  ```

## 特性

1. immutable：内部使用 cloneDeep 函数，对传入的数据进行深拷贝，所有修改不影响原始数据
   ```js
   constructor(data: any) {
    this.originalData = deepClone(data);
    this.result = deepClone(data);
   }
   ```
2. 方法调用顺序无关：可以自由使用链式调用方法，内部使用固定的方法执行顺序（add[增加数据]->delete[删除数据]->editValue[修改 value] -> editKey[修改 key]）调用对应的方法来执行，确保数据操作的正确性
   ```js
   // 源码测试用例
   test('valueDelivery', () => {
     const data = { a: 'a', b: 'b' };
     const matcher = new Matcher(data);
     matcher
       .add('c', () => 'c')
       .delete(['b'])
       .editValue('a', () => 'aa');
     expect(matcher.data).toEqual({ a: 'aa', c: 'c' }); // pass
     // 顺序无关
     matcher
       .editValue('a', () => 'aa')
       .add('c', () => 'c')
       .delete(['b']);
     expect(matcher.data).toEqual({ a: 'aa', c: 'c' }); // pass
   });
   ```
3. 归纳操作：Matcher 内部会收集所有调用方法，便于对数组数据只使用一次遍历完成数据操作
4. 链式调用：使用链式调用方法，对数据的操作代码更有组织性


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.2.3](https://github.com/xiamu14/data-matcher/compare/v2.2.2...v2.2.3) (2021-11-27)

### [2.2.2](https://github.com/xiamu14/data-matcher/compare/v2.2.1...v2.2.2) (2021-11-23)

### [2.2.1](https://github.com/xiamu14/data-matcher/compare/v2.2.0...v2.2.1) (2021-10-05)

## [2.2.0](https://github.com/xiamu14/data-matcher/compare/v2.1.8...v2.2.0) (2021-10-04)

### [2.1.8](https://github.com/xiamu14/data-matcher/compare/v2.1.7...v2.1.8) (2021-09-29)

### [2.1.7](https://github.com/xiamu14/data-matcher/compare/v2.1.6...v2.1.7) (2021-09-29)

### [2.1.6](https://github.com/xiamu14/data-matcher/compare/v2.1.5...v2.1.6) (2021-09-29)

### [2.1.5](https://github.com/xiamu14/data-matcher/compare/v2.1.4...v2.1.5) (2021-09-29)

### [2.1.4](https://github.com/xiamu14/data-matcher/compare/v2.1.3...v2.1.4) (2021-09-29)

### [2.1.3](https://github.com/xiamu14/data-matcher/compare/v2.1.2...v2.1.3) (2021-09-29)

### [2.1.2](https://github.com/xiamu14/data-matcher/compare/v2.1.1...v2.1.2) (2021-09-29)

### [2.1.1](https://github.com/xiamu14/data-matcher/compare/v2.1.0...v2.1.1) (2021-09-29)

## 2.1.0 (2021-09-29)


### Features

* add .d.ts. ([59a5a85](https://github.com/xiamu14/data-matcher/commit/59a5a8516f8837b578b163d3635b3fce519f2555))
* add contrubute. ([2378f21](https://github.com/xiamu14/data-matcher/commit/2378f212430c6d0845de16de570c545ed06eef54))
* add new feature function tranformVal. ([0993bd9](https://github.com/xiamu14/data-matcher/commit/0993bd909984772be427a63d88a19245a77bb073))
* add new feature. ([363a624](https://github.com/xiamu14/data-matcher/commit/363a62434a08fe60396cbf505f208c609462bfe6))
* add part matcher. ([c913338](https://github.com/xiamu14/data-matcher/commit/c9133384e41328694aa128c0f3d697a0ce08fe89))
* add some file. ([604dac1](https://github.com/xiamu14/data-matcher/commit/604dac1b261bae0789941bc63491a55a38a32815))
* add some new type change. ([73b764f](https://github.com/xiamu14/data-matcher/commit/73b764fafe1c538df49064c82df6a80f2bd26823))
* add somechange. ([902c42d](https://github.com/xiamu14/data-matcher/commit/902c42da82b85b51229f4513fac570ef02b254d7))
* add test instance. ([bc858b8](https://github.com/xiamu14/data-matcher/commit/bc858b8cb90399da1d6915c28448459a746e5b5b))
* adjust to support typescript. ([95e7b51](https://github.com/xiamu14/data-matcher/commit/95e7b517c3cc977e63bc979a40373861fd4cfc8f))
* all function support array and add test suit. ([5ca547c](https://github.com/xiamu14/data-matcher/commit/5ca547c90371b2e6b327537c0fdc47fda679bd71))
* change export class. ([08823f7](https://github.com/xiamu14/data-matcher/commit/08823f7100e90a7180cafe5f1f54f3ec9e368388))
* change log. ([3e97530](https://github.com/xiamu14/data-matcher/commit/3e975301bbc81256050b4513c824762e62418e01))
* change pkg name. ([9f4fc84](https://github.com/xiamu14/data-matcher/commit/9f4fc845ccb575d272dfcf9d2d45ec14992b4eb6))
* i forget what i write. ([eb256c0](https://github.com/xiamu14/data-matcher/commit/eb256c0e78068f13732d6b9975f072c9a350ff33))
* init library. ([689c0b5](https://github.com/xiamu14/data-matcher/commit/689c0b54caae98b1333a9e27caabb0d6fa05d262))
* publish new version. ([bad73e0](https://github.com/xiamu14/data-matcher/commit/bad73e084e4cfe89aa4855c0327b8ea465f84bb0))
* uglify some function. ([0363e44](https://github.com/xiamu14/data-matcher/commit/0363e44aa55b5a96b8a4caa5b994ebc7e19afae5))
* uglify, export part function. ([6d3cc79](https://github.com/xiamu14/data-matcher/commit/6d3cc79ea744fcddda752a3c54fc59fd6f943cf0))
* update readme.md. ([4afa75e](https://github.com/xiamu14/data-matcher/commit/4afa75ea689ae338213a3141626b207ba6d9cef7))
* update registry. ([23f3602](https://github.com/xiamu14/data-matcher/commit/23f3602b35d3cf6da48757f56975843d0744d568))
* update version. ([c217b33](https://github.com/xiamu14/data-matcher/commit/c217b338a4aa5d6b238ef7c7a002291f60f3d724))
* update. ([7b834ff](https://github.com/xiamu14/data-matcher/commit/7b834ffcf1084cf8b69bca762b3cfbdcaea80518))
* use old code . ([8475513](https://github.com/xiamu14/data-matcher/commit/84755133b4cf7c6902194ecdef21fb5f25bf4896))


### Bug Fixes

* fix type description error. ([fba7d11](https://github.com/xiamu14/data-matcher/commit/fba7d11fdda590d400afd55f419adb171220ae5b))


lib/index.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = global || self, factory(global['data-matcher'] = {}));
}(this, function (exports) { 'use strict';

  /*
   * @Description: 将对象转换为数组，扁平化
   * @Type: function
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-25 11:15:21
   * @LastEditTime: 2019-04-25 11:17:18
   */
  function toValuesArray(data) {
      return Object.values(data);
  }

  /*
   * @Description: 转换 key 的名字
   * @Type: function
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-25 11:17:37
   * @LastEditTime: 2019-04-25 11:19:24
   */
  function transfromKey(data, map) {
      var targetKeys = Object.keys(map);
      var target = {};
      Object.keys(data).forEach(function (key) {
          if (targetKeys.includes(key)) {
              target[map[key]] = data[key];
          }
          else {
              target[key] = data[key];
          }
      });
      return target;
  }

  /*
   * @Description:
   * @Type:
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-26 10:18:15
   * @LastEditTime: 2019-04-26 10:23:21
   */
  function transformValueType(data, map) {
      var target = {};
      Object.keys(data).forEach(function (key) {
          if (map[key] === 'number') {
              target[key] = parseInt(data[key], 10);
          }
          else if (map[key] === 'string') {
              target[key] = data[key].toString();
          }
          else {
              target[key] = data[key];
          }
      });
      return target;
  }

  /*
   * @Description:
   * @Type:
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-26 10:24:30
   * @LastEditTime: 2019-04-26 10:25:40
   */
  function transfromEmpty(data) {
      var target = {};
      Object.keys(data).forEach(function (key) {
          if (data[key] === null ||
              data[key] === 'undefined' ||
              data[key] === '"undefined"') {
              target[key] = '';
          }
          else {
              target[key] = data[key];
          }
      });
      return target;
  }

  /*
   * @Description:
   * @Type:
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-26 10:27:37
   * @LastEditTime: 2019-04-26 10:28:39
   */
  function addKey(data, map) {
      var target = data;
      Object.keys(map).forEach(function (key) {
          target[key] = map[key];
      });
      return target;
  }

  /*
   * @Description:
   * @Type:
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-26 10:33:16
   * @LastEditTime: 2019-04-26 10:39:03
   */
  function part(data, key, fn) {
      if (!Object.prototype.hasOwnProperty.bind(data, key)) {
          throw new RangeError("the data hasn't key " + key);
      }
      var target = data;
      target[key] = fn && fn(data[key]);
      return target;
  }

  /*
   * @Description:
   * @Type:
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-26 10:29:22
   * @LastEditTime: 2019-04-26 10:30:47
   */
  function addKeyFn(data, key, fn) {
      var target = data;
      target[key] = fn && fn(data);
      return target;
  }

  /*
   * @Description:根据 key 值分类，处理的是数组
   * @Type:
   * @Author: Ben
   * @LastEditors: Ben
   * @Date: 2019-04-26 10:39:22
   * @LastEditTime: 2019-04-26 10:43:12
   */
  function classifyKey(data, key) {
      var target = [];
      data.forEach(function (item) {
          var val = item[key];
          if (!Object.keys(target).includes(val.toString())) {
              target[val] = [];
          }
          target[val].push(item);
      });
      return target;
  }

  exports.toValuesArray = toValuesArray;
  exports.transformKey = transfromKey;
  exports.transformValueType = transformValueType;
  exports.transformEmpty = transfromEmpty;
  exports.addKey = addKey;
  exports.part = part;
  exports.addKeyFn = addKeyFn;
  exports.classifyKey = classifyKey;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
